pkg/tools/cli/loadtest/service/loader.go

```diff
@@ -88,48 +99,64 @@ func (l *loader) Prepare(context.Context) (err error) {
 	if err != nil {
 		return err
 	}
-	l.requests, err = l.requestsFunc(dataset)
+
+	switch l.operation {
+	case config.Insert, config.StreamInsert:
+		l.dataProvider, l.dataSize, err = insertRequestProvider(dataset, l.batchSize)
+	case config.Search, config.StreamSearch:
+		l.dataProvider, l.dataSize, err = searchRequestProvider(dataset)
+	}
 	if err != nil {
 		return err
 	}
 
-	if len(l.requests) == 0 {
-		return errors.New("prepare data is empty")
-	}
-
 	return nil
 }
 
 // Do operates load test.
 func (l *loader) Do(ctx context.Context) <-chan error {
-	ech := make(chan error, len(l.requests))
-
-	// TODO: related to #403.
-	p, err := os.FindProcess(os.Getpid())
-	if err != nil {
+	ech := make(chan error, l.dataSize)
+	finalize := func(ctx context.Context, err error) {
 		select {
 		case <-ctx.Done():
 			ech <- errors.Wrap(err, ctx.Err().Error())
 		case ech <- err:
 		}
+	}
+
+	// TODO: related to #403.
+	p, err := os.FindProcess(os.Getpid())
+	if err != nil {
+		finalize(ctx, err)
 		return ech
 	}
 
-	var pgCnt int32 = 0
-	var start time.Time
+	var pgCnt, errCnt int32 = 0, 0
+	var start, end time.Time
+	vps := func(count int, t1, t2 time.Time) float64 {
+		return float64(count) / t2.Sub(t1).Seconds()
+	}
 	progress := func() {
-		log.Infof("progress %d items, %f[qps]", pgCnt, float64(pgCnt)/time.Now().Sub(start).Seconds())
+		log.Infof("progress %d requests, %f[vps], error: %d", pgCnt, vps(int(pgCnt)*l.batchSize, start, time.Now()), errCnt)
 	}
+
+	f := func(i interface{}, err error) {
+		atomic.AddInt32(&pgCnt, 1)
+		if err != nil {
+			atomic.AddInt32(&errCnt, 1)
+		}
+	}
+
+	ticker := time.NewTicker(l.progressDuration)
 	l.eg.Go(safety.RecoverFunc(func() error {
-		ticker := time.NewTicker(l.progressDuration)
 		defer ticker.Stop()
-		for pgCnt != int32(len(l.requests)) {
+		for int(pgCnt) < l.dataSize {
```

Comment:
sorry :bow:
```suggestion
		for int(atomic.LoadInt32(pbCnt)) < l.dataSize {
```
pgCount is shared by multiple goroutine, so I feel atomic.LoadInt32 () is better.